### PR TITLE
Server side analytics

### PR DESCRIPTION
**Pull request by davidfischer**
This PR sends an [event](https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#event) to Google Analytics from the server side when the Footer API is called.

This is one of our [identified possible alternatives](https://docs.readthedocs.io/en/latest/advertising-details.html#alternatives) to client side GA. Sending data from the server side has many advantages such as being able to anonymize data *before* it is sent and not having clients directly connect to a site they may not want to. It comes with the downside of needing to have many web servers to handle all our analytics calls.

I am seeking feedback on the structure especially around the structure in terms of using signals/hooks to find when footer calls are made and whether this seems like a sane approach.

### CODE REVIEW

**Comment by davidfischer** [comment]
I'm expecting a lint error here around the naming of the `log` variable but I believe that means our linting is a bit overzealous.

File: readthedocs/analytics/models.py [hunk 1]
```diff
@@ -0,0 +1 @@
+"""Intentionally blank"""
```
Comment (ericholscher) [inline readthedocs/analytics/models.py]: I don't think this is necessary with the AppConfig.

File: readthedocs/analytics/signals.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    data = {
+        'ec': 'footer-api',
+        'ea': 'load',
+        'el': project.slug,
+
+        # User data
+        'ua': request.META.get('HTTP_USER_AGENT'),
+        'uip': get_client_ip(request),
+    }
+
```
Comment (ericholscher) [inline readthedocs/analytics/signals.py]: It doesn't look like this is gated by the feature flag, so it would start sending _all_ footer data.

Comment (davidfischer) [reply readthedocs/analytics/signals.py]: Correct, it is not yet.

**Comment by davidfischer** [comment]
> For this use case, I think it might actually be more useful to have a sample that isn't project specific

How about sending ad clicks to GA server side rather than client side? That would be hundreds (but not thousands) of clicks per day.

File: readthedocs/analytics/tasks.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    :see: https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#event
+    :param event_data: event parameters to send to GA
+    """
+    data = {
+        't': 'event',       # GA event - don't change
+        'ec': None,         # Event category (required)
+        'ea': None,         # Event action (required)
+        'el': None,         # Event label
+        'ev': None,         # Event value (numeric)
```
Comment (ericholscher) [inline readthedocs/analytics/tasks.py]: Should we be setting these to None by default? It seems like we should be a bit more defensive here, and make sure the incoming `event_data` contains them.

File: readthedocs/analytics/utils.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+try:
+    # Python 3.3+ only
+    import ipaddress
+except ImportError:
+    from .vendor import ipaddress
+
+log = logging.getLogger(__name__)   # noqa
+
+# Used to anonymize an IP by zero-ing out the last 2 bytes
+MASK = int('0xFFFFFFFFFFFFFFFFFFFFFFFFFFFF0000', 16)
```
Comment (ericholscher) [inline readthedocs/analytics/utils.py]: Should this just live in the `anonymize_ip_address` function? Not sure if we need it elsewhere.

File: readthedocs/analytics/utils.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        # Anonymize user agent if it is rare
+        data['ua'] = anonymize_user_agent(data['ua'])
+
+    resp = None
+    try:
+        resp = requests.post(
+            'https://www.google-analytics.com/collect',
+            data=data,
+        )
+    except requests.Timeout:
```
Comment (ericholscher) [inline readthedocs/analytics/utils.py]: What is the default timeout here? We should probably set it to something really low.

File: readthedocs/core/static-src/core/js/doc-embed/sponsorship.js [hunk 6]
```diff
@@ -87,29 +87,13 @@ function Promo(data) {
...
-        // as well as the newer universal analytics
-        if (typeof ga !== 'undefined') {
-            ga('rtfd.send', 'event', 'Promo', 'Click', data.id);
-        } else if (typeof _gaq !== 'undefined') {
-            _gaq.push(
-                ['rtfd._setAccount', 'UA-17997319-1'],
-                ['rtfd._trackEvent', 'Promo', 'Click', data.id]
-            );
-        }
-    };
```
Comment (ericholscher) [inline readthedocs/core/static-src/core/js/doc-embed/sponsorship.js]: I wonder about leaving this logic in, until we can properly confirm the server-side stuff is working. Not sure how important it is, if it's better to possible double count or to just lose the data.

**Comment by davidfischer** [comment]
I made the updates based on the feedback including making it more obvious what is required in terms of analytics.

**Comment by ericholscher** [comment]
💯

**Pull request merged**